bundles/org.openhab.binding.mpd/src/main/java/org/openhab/binding/mpd/internal/protocol/MPDConnectionThread.java

```diff
@@ -0,0 +1,298 @@
+            try {
+                socket.close();
+            } catch (IOException e) {
+            }
+            this.socket = null;
+        }
+    }
+
+    /**
+     * add a command to the pending commands queue
+     *
+     * @param command command to add
+     */
+    public void addCommand(MPDCommand command) {
+        insertCommand(command, -1);
+    }
+
+    private void insertCommand(MPDCommand command, int position) {
+        logger.debug("insert command '{}' at position {}", command.getCommand(), position);
+        int index = position;
+        synchronized (pendingCommands) {
+            if (index < 0) {
+                index = pendingCommands.size();
+            }
+            pendingCommands.add(index, command);
+            sendNoIdleIfInIdle();
+        }
+    }
+
+    private void updateThingStatus(ThingStatus status, ThingStatusDetail statusDetail, @Nullable String description) {
+        if (!disposed.get()) {
+            listener.updateThingStatus(status, statusDetail, description);
+        }
+    }
+
+    private void sendNoIdleIfInIdle() {
+        if (isInIdle.compareAndSet(true, false)) {
+            try {
+                sendCommand(new MPDCommand("noidle"));
+            } catch (IOException e) {
+                logger.debug("sendCommand(noidle) failed");
+            }
+        }
+    }
+
+    private void establishConnection() throws IOException, MPDException {
+        openSocket();
+
+        MPDCommand currentCommand = new MPDCommand("connect");
+        MPDResponse response = readResponse(currentCommand);
+
+        if (!response.isOk()) {
+            throw new MPDException("could not connect");
+        }
+
+        if (!password.isEmpty()) {
+            currentCommand = new MPDCommand("password", password);
+            sendCommand(currentCommand);
+            response = readResponse(currentCommand);
+            if (!response.isOk()) {
+                throw new MPDException("could not authenticate");
+            }
+        }
+    }
+
+    private void openSocket() throws IOException {
+        logger.debug("opening connection to {} port {}", address, port);
+        Socket socket = new Socket(address, port);
+
+        inputStreamReader = new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8);
+        reader = new BufferedReader(inputStreamReader);
+        writer = new DataOutputStream(socket.getOutputStream());
+
+        this.socket = socket;
+    }
+
+    private void processPendingCommands() throws IOException {
+        MPDCommand currentCommand;
+
+        while (!disposed.get()) {
+            synchronized (pendingCommands) {
+                if (!pendingCommands.isEmpty()) {
+                    currentCommand = pendingCommands.remove(0);
+                } else {
+                    currentCommand = new MPDCommand("idle");
+                }
+
+                sendCommand(currentCommand);
+                if (currentCommand.getCommand().equals("idle")) {
```

Comment:
Position literals first in `String` comparisons - that way if the `String` is `null` you won't get a `NullPointerException`, it'll just return `false`. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)